File: rust/ccwc.rs
use std::env;
use std::process;

mod file_reader;
use file_reader::FileReader;

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 || args.len() > 3 {
        eprintln!(
            "Usage: {} [OPTION] <file_path> \n
            \n
            OPTIONS:\n
                -c, --bytes\n
                    print the byte counts\n
                -l, --lines\n
                    print the newline counts\n
            ",
            args[0]
        );
        process::exit(1);
    }

    let mode = get_mode(&args);
    let file_path = &args[args.len() -1];

    let file_reader = FileReader::new(String::from(file_path));
    process_file(mode, &file_reader);    
}

fn process_file(mode: &str, file_reader: &FileReader) {
    match mode {
        "all" => print_all_info(file_reader),
        "bytes" => print_bytes(file_reader),
        "lines" => print_lines(file_reader),
        "" => print_all_info(file_reader),
        _ => unreachable!(),
    }
}

fn get_mode(args: &[String]) -> &str {
    if args.len() == 3 {
        match args[1].as_str() {
            "-c" | "--bytes" => "bytes",
            "-l" | "--lines" => "lines",
            _ => "unknown",
        }
    } else {
        "all"
    }
}

fn print_all_info(file_reader: &FileReader) {
    let Ok(byte_count) = file_reader.count_bytes() else { 
        panic!("Error reading size of file for '{}'", file_reader.file_path)
    };
    match file_reader.count_lines_and_words() {
        Ok((line_count, word_count)) => {
              println!(
                "{} {} {} {}",
                line_count,
                word_count,
                byte_count,
                file_reader.file_path
            );
        }
        Err(e) => {
            eprintln!("Error reading metadata for '{}': {}", file_reader.file_path, e);
            process::exit(1);
        }
    }
}

fn print_bytes(file_reader: &FileReader) {
    match file_reader.count_bytes() {
        Ok(byte_count) => {
            println!("{} {}", byte_count, file_reader.file_path);
        }
        Err(e) => {
            eprintln!("Error reading metadata for '{}': {}", file_reader.file_path, e);
            process::exit(1);
        }
    }
}

fn print_lines(file_reader: &FileReader) {
    match file_reader.count_lines() {
        Ok(line_count) => {
            println!("{} {}", line_count, file_reader.file_path);
        }
        Err(e) => {
            eprintln!("Error reading metadata for '{}': {}", file_reader.file_path, e);
            process::exit(1);
        }
    }
}

File: cpp/ccwc.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

using namespace std;

int getWordCount(std::ifstream& inputFile)
{
    string line;
    int wordCount = 0;
    while (getline(inputFile, line)) {
        std::istringstream iss(line);
        std::string word;
        while (iss >> word) {
            wordCount++;
        }
    }
    return wordCount;
}

int getLineCount(std::ifstream& inputFile)
{
    string line;
    int lineCount = 0;

    while (getline(inputFile, line)) {
        ++lineCount;
    }
    return lineCount;
}

pair<int, int> getWordAndLineCount(std::ifstream& inputFile)
{
    string line;
    int lineCount = 0;
    int wordCount = 0;

    while (getline(inputFile, line)) {
        ++lineCount;
        std::istringstream iss(line);
        std::string word;
        while (iss >> word) {
            wordCount++;
        }
    }
    return { wordCount, lineCount };
}

void printAll(std::ifstream& inputFile, const string& filePath)
{
    inputFile.seekg(0, ios::end);
    streampos fileSize = inputFile.tellg();
    inputFile.seekg(0, ios::beg);

    auto [wordCount, lineCount] = getWordAndLineCount(inputFile);
    cout << lineCount << " " << wordCount << " " << fileSize << " " << filePath << endl;
}

void printBytes(std::ifstream& inputFile, const string& filePath)
{
    inputFile.seekg(0, ios::end);
    streampos fileSize = inputFile.tellg();
    inputFile.seekg(0, ios::beg);

    cout << fileSize << " " << filePath << endl;
}

void printLines(std::ifstream& inputFile, const string& filePath)
{
    int lineCount = getLineCount(inputFile);
    cout << lineCount << " " << filePath << endl;
}

void printWords(std::ifstream& inputFile, const string& filePath)
{
    int wordCount = getWordCount(inputFile);
    cout << wordCount << " " << filePath << endl;
}

int processFile(const string& option, const string& filePath)
{
    ifstream inputFile(filePath);

    if (!inputFile.is_open()) {
        cerr << "Error opening the file!" << endl;
        return 1;
    }

    try {
        if (option == "-c" || option == "--bytes") {
            printBytes(inputFile, filePath);
        } else if (option == "-l" || option == "--lines") {
            printLines(inputFile, filePath);
        } else if (option == "-a") {
            printAll(inputFile, filePath);
        } else if (option == "-w" || option == "--words") {
            printWords(inputFile, filePath);
        } else {
            cerr << "Unkown flag " << option << endl;
        }

        return 0;
    } catch (const std::ifstream::failure& e) {
        cerr << "Error reading file " << e.what() << endl;
        return 1;
    }
}

int main(int argc, char* argv[])
{
    if (argc < 2 || argc > 3) {
        cerr << "Usage: " << argv[0] << " [OPTION] <filePath>\n"
             << "\n"
             << "OPTIONS:\n"
             << "   -c, --bytes\n"
             << "       print the byte counts\n"
             << "   -l, --lines\n"
             << "       print the newline counts\n"
             << "   -w, --words\n"
             << "       print the word counts\n"
             << endl;
        return 1;
    }

    string option = (argc == 3) ? argv[1] : "-a";
    string filePath = argv[argc - 1];
    processFile(option, filePath);

    return 0;
}
